File: AzanAlarm-RN/src/services/LocationService.ts
import Geolocation from '@react-native-community/geolocation';
import axios from 'axios';
import { Location } from '../types';
import { storageService } from './StorageService';
import { STORAGE_KEYS, API_ENDPOINTS } from '../constants/storage';
import { LocationError, NetworkError } from '../utils/ErrorHandler';
import { checkLocationPermission, requestLocationPermission, PermissionStatus } from '../utils/PermissionUtils';

export interface LocationService {
    getCurrentLocation(): Promise<Location | null>;
    searchLocations(query: string): Promise<Location[]>;
    saveLocation(location: Location): Promise<void>;
    getSavedLocations(): Promise<Location[]>;
    setCurrentLocation(location: Location): Promise<void>;
}

interface NominatimResult {
    place_id: number;
    lat: string;
    lon: string;
    display_name: string;
    address: {
        country?: string;
        city?: string;
        town?: string;
        village?: string;
    };
}

class LocationServiceImpl implements LocationService {
    /**
     * Get current location using GPS
     */
    async getCurrentLocation(): Promise<Location | null> {
        try {
            // Check permission first
            let permissionStatus = await checkLocationPermission();

            if (permissionStatus !== PermissionStatus.GRANTED) {
                permissionStatus = await requestLocationPermission();

                if (permissionStatus !== PermissionStatus.GRANTED) {
                    throw new LocationError(
                        'Location permission denied',
                        'PERMISSION_DENIED'
                    );
                }
            }

            // Get current position
            return new Promise((resolve, reject) => {
                Geolocation.getCurrentPosition(
                    async (position) => {
                        try {
                            const { latitude, longitude } = position.coords;

                            // Reverse geocode to get location name
                            const response = await axios.get<NominatimResult>(
                                API_ENDPOINTS.NOMINATIM_REVERSE,
                                {
                                    params: {
                                        lat: latitude,
                                        lon: longitude,
                                        format: 'json',
                                    },
                                    headers: {
                                        'User-Agent': 'AzanAlarm/1.0',
                                    },
                                }
                            );

                            const data = response.data;
                            const locationName = data.address.city ||
                                data.address.town ||
                                data.address.village ||
                                'Unknown';

                            const location: Location = {
                                name: locationName,
                                country: data.address.country || 'Unknown',
                                latitude,
                                longitude,
                                timezone: Intl.DateTimeFormat().resolvedOptions().timeZone,
                                isCurrent: true,
                                createdAt: new Date().toISOString(),
                            };

                            // Save as current location
                            await this.setCurrentLocation(location);

                            resolve(location);
                        } catch (error) {
                            reject(new NetworkError(
                                'Failed to reverse geocode location',
                                'GEOCODING_ERROR',
                                error as Error
                            ));
                        }
                    },
                    (error) => {
                        reject(new LocationError(
                            'Failed to get current location',
                            'GPS_ERROR',
                            error as any
                        ));
                    },
                    {
                        enableHighAccuracy: true,
                        timeout: 15000,
                        maximumAge: 10000,
                    }
                );
            });
        } catch (error) {
            console.error('Failed to get current location:', error);
            throw error;
        }
    }

    /**
     * Search for locations using Nominatim
     */
    async searchLocations(query: string): Promise<Location[]> {
        try {
            if (!query || query.trim().length < 2) {
                return [];
            }

            const response = await axios.get<NominatimResult[]>(
                API_ENDPOINTS.NOMINATIM_SEARCH,
                {
                    params: {
                        q: query,
                        format: 'json',
                        limit: 10,
                        addressdetails: 1,
                    },
                    headers: {
                        'User-Agent': 'AzanAlarm/1.0',
                    },
                }
            );

            return response.data.map((result) => ({
                name: result.address.city ||
                    result.address.town ||
                    result.address.village ||
                    result.display_name.split(',')[0],
                country: result.address.country || 'Unknown',
                latitude: parseFloat(result.lat),
                longitude: parseFloat(result.lon),
                timezone: Intl.DateTimeFormat().resolvedOptions().timeZone,
                createdAt: new Date().toISOString(),
            }));
        } catch (error) {
            throw new NetworkError(
                'Failed to search locations',
                'SEARCH_ERROR',
                error as Error
            );
        }
    }

    /**
     * Save a location to favorites
     */
    async saveLocation(location: Location): Promise<void> {
        try {
            const savedLocations = await this.getSavedLocations();

            // Check if location already exists
            const exists = savedLocations.some(
                (loc) => loc.latitude === location.latitude && loc.longitude === location.longitude
            );

            if (!exists) {
                const newLocation = {
                    ...location,
                    id: Date.now(),
                    createdAt: new Date().toISOString(),
                };

                savedLocations.push(newLocation);
                storageService.set(STORAGE_KEYS.SAVED_LOCATIONS, savedLocations);
            }
        } catch (error) {
            console.error('Failed to save location:', error);
            throw error;
        }
    }

    /**
     * Get all saved locations
     */
    async getSavedLocations(): Promise<Location[]> {
        try {
            const locations = storageService.get<Location[]>(STORAGE_KEYS.SAVED_LOCATIONS);
            return locations || [];
        } catch (error) {
            console.error('Failed to get saved locations:', error);
            return [];
        }
    }

    /**
     * Set current active location
     */
    async setCurrentLocation(location: Location): Promise<void> {
        try {
            const currentLocation = {
                ...location,
                isCurrent: true,
            };

            storageService.set(STORAGE_KEYS.CURRENT_LOCATION, currentLocation);
        } catch (error) {
            console.error('Failed to set current location:', error);
            throw error;
        }
    }
}

export const locationService = new LocationServiceImpl();

File: internal/services/location_service.go
// Package services contains business logic for the AzanAlarm application.
package services

import (
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"time"

	"AzanAlarm/internal/models"
)

// LocationService handles location-related operations.
type LocationService struct {
	storage    *StorageService
	httpClient *http.Client
}

// NominatimResult represents a result from the Nominatim API.
type NominatimResult struct {
	PlaceID     int    `json:"place_id"`
	Lat         string `json:"lat"`
	Lon         string `json:"lon"`
	DisplayName string `json:"display_name"`
	Address     struct {
		Country string `json:"country"`
		City    string `json:"city"`
		Town    string `json:"town"`
		Village string `json:"village"`
	} `json:"address"`
}

// NewLocationService creates a new LocationService instance.
func NewLocationService(storage *StorageService) *LocationService {
	return &LocationService{
		storage: storage,
		httpClient: &http.Client{
			Timeout: 15 * time.Second,
		},
	}
}

// SearchLocations searches for locations using the Nominatim API.
func (ls *LocationService) SearchLocations(query string) ([]models.Location, error) {
	if len(query) < 2 {
		return []models.Location{}, nil
	}

	baseURL := "https://nominatim.openstreetmap.org/search"
	params := url.Values{}
	params.Add("q", query)
	params.Add("format", "json")
	params.Add("limit", "10")
	params.Add("addressdetails", "1")

	req, err := http.NewRequest("GET", baseURL+"?"+params.Encode(), nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("User-Agent", "AzanAlarm/1.0")

	resp, err := ls.httpClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	var results []NominatimResult
	if err := json.NewDecoder(resp.Body).Decode(&results); err != nil {
		return nil, err
	}

	locations := make([]models.Location, 0, len(results))
	for _, r := range results {
		lat, _ := parseFloat(r.Lat)
		lon, _ := parseFloat(r.Lon)

		name := r.Address.City
		if name == "" {
			name = r.Address.Town
		}
		if name == "" {
			name = r.Address.Village
		}
		if name == "" && r.DisplayName != "" {
			// Use first part of display name
			name = r.DisplayName
			if len(name) > 30 {
				name = name[:30]
			}
		}

		locations = append(locations, models.Location{
			Name:      name,
			Country:   r.Address.Country,
			Latitude:  lat,
			Longitude: lon,
			Timezone:  time.Local.String(), // Approximate
			CreatedAt: time.Now().UnixMilli(),
		})
	}

	return locations, nil
}

// ReverseGeocode gets location name from coordinates.
func (ls *LocationService) ReverseGeocode(lat, lon float64) (*models.Location, error) {
	baseURL := "https://nominatim.openstreetmap.org/reverse"
	params := url.Values{}
	params.Add("lat", formatFloat(lat))
	params.Add("lon", formatFloat(lon))
	params.Add("format", "json")

	req, err := http.NewRequest("GET", baseURL+"?"+params.Encode(), nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("User-Agent", "AzanAlarm/1.0")

	resp, err := ls.httpClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	var result NominatimResult
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, err
	}

	name := result.Address.City
	if name == "" {
		name = result.Address.Town
	}
	if name == "" {
		name = result.Address.Village
	}
	if name == "" {
		name = "Unknown"
	}

	return &models.Location{
		Name:      name,
		Country:   result.Address.Country,
		Latitude:  lat,
		Longitude: lon,
		Timezone:  time.Local.String(),
		IsCurrent: true,
		CreatedAt: time.Now().UnixMilli(),
	}, nil
}

// GetCurrentLocation returns the currently active location.
func (ls *LocationService) GetCurrentLocation() *models.Location {
	var location models.Location
	if err := ls.storage.Load("current_location", &location); err != nil {
		return nil
	}
	return &location
}

// SetCurrentLocation sets the current active location.
func (ls *LocationService) SetCurrentLocation(location models.Location) error {
	location.IsCurrent = true
	return ls.storage.Save("current_location", location)
}

// GetSavedLocations returns all saved locations.
func (ls *LocationService) GetSavedLocations() []models.Location {
	var locations []models.Location
	if err := ls.storage.Load("saved_locations", &locations); err != nil {
		return []models.Location{}
	}
	return locations
}

// SaveLocation adds a location to saved locations.
func (ls *LocationService) SaveLocation(location models.Location) error {
	locations := ls.GetSavedLocations()

	// Check if location already exists
	for _, loc := range locations {
		if loc.IsSameLocation(location) {
			return nil // Already saved
		}
	}

	location.ID = len(locations) + 1
	locations = append(locations, location)
	return ls.storage.Save("saved_locations", locations)
}

// DeleteLocation removes a location from saved locations.
func (ls *LocationService) DeleteLocation(id int) error {
	locations := ls.GetSavedLocations()
	newLocations := make([]models.Location, 0, len(locations))

	for _, loc := range locations {
		if loc.ID != id {
			newLocations = append(newLocations, loc)
		}
	}

	return ls.storage.Save("saved_locations", newLocations)
}

// formatFloat formats a float64 for URL parameters.
func formatFloat(f float64) string {
	return fmt.Sprintf("%f", f)
}

// parseFloat parses a string to float64.
func parseFloat(s string) (float64, error) {
	var f float64
	_, err := fmt.Sscanf(s, "%f", &f)
	return f, err
}
